python/modules/tauMVA.py

```diff
@@ -0,0 +1,137 @@
+			maxPhotonPT = c.pt;
+			photonInd = ic;
+	
+	return photonInd;
+
+    def transverseMass(self, visible, invisible):
+	cosDPhi   = np.cos( deltaPhi(visible.Phi(), invisible.phi) );
+	return np.sqrt( 2 * visible.Pt() * invisible.pt * (1 - cosDPhi) );
+        
+    def computeMT(self, pfc, met, pfcands):
+	photonInd = self.getNearPhotonIndex(pfc, pfcands);
+	candP4 = ROOT.TLorentzVector()
+	candP4.SetPtEtaPhiM(pfc.pt, pfc.eta, pfc.phi, pfc.mass)
+	if(photonInd > -1): 
+		pfcand_buff = ROOT.TLorentzVector()
+		pfcand_buff.SetPtEtaPhiM(pfcands[photonInd].pt, pfcands[photonInd].eta, pfcands[photonInd].phi, pfcands[photonInd].mass)
+		candP4+=pfcand_buff;
+	return self.transverseMass(candP4, met);
+
+    def analyze(self, event):
+        ## Getting objects
+	met	  = Object(event, self.metBranchName)
+	jets	  = Collection(event, "Jet")
+	pfcand    = Collection(event, "PFcand")
+	eventNum  = event.event
+
+        pfchargedhads = []
+	mva = {}
+	mva_ = []
+	mt_ = []
+	for pfc in pfcand:
+		mva_buff = 0.0
+		mt = 0.0
+		if(pfc.pt > 10.0 and abs(pfc.eta) < 2.4 and abs(pfc.dz) < 0.2):
+			mt = self.computeMT(pfc, met, pfcand)
+
+			pt 	     = min(pfc.pt,float(300.0))
+			abseta       = min(abs(pfc.eta), float(2.4))
+			chiso0p1     = min(pfc.chiso0p1,float(700.0))
+			chiso0p2     = min(pfc.chiso0p2,float(700.0))
+			chiso0p3     = min(pfc.chiso0p3,float(700.0))
+			chiso0p4     = min(pfc.chiso0p4,float(700.0))
+			totiso0p1    = min(pfc.totiso0p1,float(700.0))
+			totiso0p2    = min(pfc.totiso0p2,float(700.0))
+			totiso0p3    = min(pfc.totiso0p3,float(700.0))
+			totiso0p4    = min(pfc.totiso0p4,float(700.0))
+			neartrkdr    = pfc.nearestTrkDR
+			jetmatch     = (pfc.contJetIndex > -1) and (jets[pfc.contJetIndex].pt >= 20.0) and (abs(jets[pfc.contJetIndex].eta) < 2.4)
+			jetdr        = deltaR(jets[pfc.contJetIndex].eta, jets[pfc.contJetIndex].phi, pfc.eta, pfc.phi) if jetmatch else -1.0
+			jetcsv       = jets[pfc.contJetIndex].btagDeepB if jetmatch else -1.0
+			
+			contjetdr  = min(float(0.4), jetdr)
+			if(contjetdr < 0.0): contjetdr = 0.0
+			contjetcsv =  jetcsv
+			if(contjetcsv < 0.0): contjetcsv = 0.0
+	
+			mva = {self.bdt_vars[0]: pt, 
+			       self.bdt_vars[1]: abseta,
+			       self.bdt_vars[2]: chiso0p1, 
+			       self.bdt_vars[3]: chiso0p2, 
+			       self.bdt_vars[4]: chiso0p3, 
+			       self.bdt_vars[5]: chiso0p4, 
+			       self.bdt_vars[6]: totiso0p1, 
+			       self.bdt_vars[7]: totiso0p2, 
+			       self.bdt_vars[8]: totiso0p3, 
+			       self.bdt_vars[9]: totiso0p4, 
+			       self.bdt_vars[10]: neartrkdr, 
+			       self.bdt_vars[11]: contjetdr, 
+			       self.bdt_vars[12]: contjetcsv}
+			mva_buff = self.xgb.eval(mva)
+		mt_.append(mt)
+		mva_.append(mva_buff)
+	self.TauMVA_Stop0l = map(self.SelTauMVA, mva_)
+
+	#print "mva output: ", mva_
+	self.out.fillBranch("mt", 		mt_)
+        self.out.fillBranch("taumva", 		mva_)
+	self.out.fillBranch("TauMVA_Stop0l", sum(self.TauMVA_Stop0l))
```

Comment:
The variable names are confusing, can you clarify?

Also, after the post-processing, I think we will drop the PFcands no? I don't see the store of the selected PFCand.

Reply:
I updated variable names and yes we can drop the PFcand variables since I am saving the BDT output

Reply:
But what about the selected PFCand pt ,eta, phi etc? I am a bit lost

Reply:
It can use the variables in the preprocessed file without storing them in the output file.

Reply:
OK, for the variables in the preprocessed file, we store lots of PFCand etc. After the post-processing, we would like to only store the selected PFCand. But I can't find out how they are reduced.

Reply:
You can see the cut on the pfcand above on line 91

Reply:
Yes, those are the cut apply to the PFCand loaded to memory, but it won't change the output. Keep in mind in the current keep_and_drop.txt, we are dropping all the PFCand. So to store the selected PFCand, we can either like createing new HOT object as in the DeepTopProducer.py, or filter the PFCand and store them as in the https://github.com/susy2015/NanoSUSY-tools/blob/dev_v2/python/modules/GenPartFilter.py

Reply:
This is just a quick attempt at trying to save the PFcand variables like to HOT variables. I couldn't get it to work quite right.